File: app/src/main/java/com/elegion/android/ui/features/FeaturesPresenter.java
package com.elegion.android.ui.features;

import android.support.annotation.NonNull;

import com.arellomobile.mvp.InjectViewState;
import com.elegion.android.data.Repository;
import com.elegion.android.data.model.Feature;
import com.elegion.android.ui.base.presenter.BasePresenter;
import com.elegion.android.util.AuthUtils;
import com.elegion.android.util.RxUtils;

import java.util.List;

import rx.Subscription;

/**
 * @author mikhail.barannikov on 24.07.2017
 */
@InjectViewState
public class FeaturesPresenter extends BasePresenter<FeaturesView> {
    private static final int PAGE_COUNT = 20;

    private int mOffset = 0;
    private Repository mRepository;
    private Subscription mLoadFeaturesSubscription;
    private boolean mIsLastPage;

    public FeaturesPresenter(@NonNull Repository repository) {
        mRepository = repository;
    }

    @Override
    protected void onFirstViewAttach() {
        loadFeatures(true);
    }

    public void loadFeatures(boolean refresh) {
        if (refresh) {
            mOffset = 0;
            mIsLastPage = false;
        }

        if (!mIsLastPage && RxUtils.isNullOrUnsubscribed(mLoadFeaturesSubscription)) {
            removeSubscription(mLoadFeaturesSubscription);
            mLoadFeaturesSubscription = mRepository.getFeatures(mOffset, PAGE_COUNT)
                    .compose(RxUtils::async)
                    .compose(RxUtils.loading(getViewState()))
                    .subscribe(this::handleFeaturesResponse, RxUtils::errorLogE);
            addSubscription(mLoadFeaturesSubscription);
        }
    }

    private void handleFeaturesResponse(List<Feature> features) {
        final boolean clear = mOffset == 0;
        if (clear) {
            getViewState().clearFeatures();
        }
        getViewState().showFeatures(features, clear);
        mOffset = mOffset + features.size();
        mIsLastPage = mOffset > 50;
    }

    public void logout() {
        AuthUtils.logout(mRepository);
    }
}

File: app/src/main/java/com/elegion/android/template/ui/features/FeaturesPresenter.kt
package com.elegion.android.template.ui.features

import com.arellomobile.mvp.InjectViewState
import com.elegion.android.template.data.Repository
import com.elegion.android.template.data.model.Feature
import com.elegion.android.template.ui.base.presenter.BasePresenter
import com.elegion.android.template.util.AuthUtils
import com.elegion.android.template.util.RxUtils
import io.reactivex.disposables.Disposable

@InjectViewState
internal class FeaturesPresenter(private val mRepository: Repository) : BasePresenter<FeaturesView>() {

    private var mOffset = 0
    private var mLoadFeaturesSubscription: Disposable? = null
    private var mIsLastPage: Boolean = false

    override fun onFirstViewAttach() = loadFeatures(true)

    fun loadFeatures(refresh: Boolean) {
        if (refresh) {
            mOffset = 0
            mIsLastPage = false
        }

        if (!mIsLastPage && RxUtils.isNullOrUnsubscribed(mLoadFeaturesSubscription)) {
            removeDisposable(mLoadFeaturesSubscription!!)
            mLoadFeaturesSubscription = mRepository.getFeatures(mOffset, PAGE_COUNT)
                    .compose<List<Feature>>({ RxUtils.async(it) })
                    .compose(RxUtils.loading(viewState))
                    .subscribe({ this.handleFeaturesResponse(it) }, { RxUtils.errorLogE(it) })
            addDisposable(mLoadFeaturesSubscription!!)
        }
    }

    private fun handleFeaturesResponse(features: List<Feature>) {
        val clear = mOffset == 0
        if (clear) {
            viewState.clearFeatures()
        }
        viewState.showFeatures(features, clear)
        mOffset += features.size
        mIsLastPage = mOffset > 50
    }

    fun logout() = AuthUtils.logout(mRepository)

    companion object {
        private const val PAGE_COUNT = 20
    }
}
